vault/resource_managed_keys.go

```diff
@@ -0,0 +1,655 @@
+	for k := range oldKeySet {
+		if !newKeySet[k] {
+			// Delete single key type
+			if diags := deleteSingleManagedKey(d, meta, keyType, k); diags != nil {
+				return diags
+			}
+		}
+	}
+
+	return nil
+}
+
+func createUpdateManagedKeys(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if _, ok := d.GetOk(consts.FieldAWS); ok {
+		if diags := writeManagedKeysData(d, meta, consts.FieldAWS); diags != nil {
+			return diags
+		}
+	}
+
+	if _, ok := d.GetOk(consts.FieldPKCS); ok {
+		if diags := writeManagedKeysData(d, meta, consts.FieldPKCS); diags != nil {
+			return diags
+		}
+	}
+
+	if _, ok := d.GetOk(consts.FieldAzure); ok {
+		if diags := writeManagedKeysData(d, meta, consts.FieldAzure); diags != nil {
+			return diags
+		}
+	}
+
+	// set ID to 'default'
+	d.SetId("default")
+
+	return readManagedKeys(ctx, d, meta)
+}
+
+func getRedactedFields(d *schema.ResourceData, providerType, name string, fields []string) map[string]interface{} {
+	m := map[string]interface{}{}
+
+	if mKeysBlocks, ok := d.GetOk(providerType); ok {
+		for _, block := range mKeysBlocks.(*schema.Set).List() {
+			data := block.(map[string]interface{})
+			if data[consts.FieldName] == name {
+				for _, field := range fields {
+					if v, ok := data[field]; ok {
+						m[field] = v
+					}
+				}
+			}
+		}
+	}
+	return m
+}
+
+func readAndSetManagedKeys(d *schema.ResourceData, meta interface{}, providerType string,
+	sm map[string]interface{}, redacted []string) error {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return e
+	}
+
+	var keyType string
+	switch providerType {
+	case consts.FieldAWS:
+		keyType = kmsTypeAWS
+
+	case consts.FieldPKCS:
+		keyType = kmsTypePKCS
+
+	case consts.FieldAzure:
+		keyType = kmsTypeAzure
+	}
+
+	p := fmt.Sprintf("%s/%s", "sys/managed-keys", keyType)
+	log.Printf("[DEBUG] Listing data from Vault at %s", p)
+	resp, err := client.Logical().List(p)
+	if err != nil {
+		return err
+	}
+
+	if resp == nil {
+		return nil
+	}
+
+	handlers := map[string]func() schemaMap{
+		kmsTypeAWS:   managedKeysAWSConfigSchema,
+		kmsTypePKCS:  managedKeysPKCSConfigSchema,
+		kmsTypeAzure: managedKeysAzureConfigSchema,
+	}
+
+	var data []interface{}
+	if v, ok := resp.Data["keys"]; ok {
+		for _, name := range v.([]interface{}) {
+			m := make(map[string]interface{})
+			path := getManagedKeysPath(keyType, name.(string))
+			log.Printf("[DEBUG] Reading from Vault at %s", path)
+			resp, err := client.Logical().Read(path)
+			if err != nil {
+				return err
+			}
+
+			if resp == nil {
+				continue
+			}
+
+			for k := range handlers[keyType]() {
+				// Map TF schema fields to Vault API
+				vaultKey := k
+				if v, ok := sm[k]; ok {
+					vaultKey = v.(string)
+				}
+
+				if v, ok := resp.Data[vaultKey]; ok {
+					m[k] = v
+				}
+			}
+
+			// set these from TF config since they are
+			// returned as "redacted"
+			redactedData := getRedactedFields(d, providerType, name.(string), redacted)
```

Comment:
maybe compute this before the `for` loop above, then you can `get()` the override value within a single `for` loop?

Reply:
@benashz I was a bit unsure what you meant by this part; how would we consolidate everything into one loop? Wouldn't we still need to loop over all redacted fields within the `getRedactedFields` function?